string.c

```diff
@@ -3554,19 +3554,30 @@ rb_str_casecmp_p(VALUE str1, VALUE str2)
 static VALUE
 str_casecmp_p(VALUE str1, VALUE str2)
 {
-    rb_encoding *enc;
-    VALUE folded_str1, folded_str2;
-    VALUE fold_opt = sym_fold;
-
-    enc = rb_enc_compatible(str1, str2);
-    if (!enc) {
-	return Qnil;
+    if (str1 == str2) return Qtrue;
+    if (!rb_enc_compatible(str1, str2)) {
+        return Qnil;
     }
 
-    folded_str1 = rb_str_downcase(1, &fold_opt, str1);
-    folded_str2 = rb_str_downcase(1, &fold_opt, str2);
+    if (ENC_CODERANGE(str1) == ENC_CODERANGE_7BIT && ENC_CODERANGE(str2) == ENC_CODERANGE_7BIT) {
+        long len1, len2;
+        const char *ptr1, *ptr2;
+
+        RSTRING_GETMEM(str1, ptr1, len1);
+        RSTRING_GETMEM(str2, ptr2, len2);
+
+        if (len1 != len2) {
+            return Qfalse;
+        }
+
+        return strncasecmp(ptr1, ptr2, len1) == 0 ? Qtrue : Qfalse;
```

Comment:
We can't use strncasecmp() here since it is for null-terminated strings and will stop comparison as soon as it hits a null character in the middle of the string.

Reply:
Interesting. I was wondering why `strncasecmp` wasn't used before. I can update with a manually crafter loop.

Reply:
Done. It now use a simple for loop and `TOLOWER()`. I see that `str_casemp()` apply ` & 0xff` to the result of `TOLOWER()`, not quite sure why, I assume it is to handle `ISO-8859-*` encoding by ignoring extended characters?